### PR TITLE
feat: emoji

### PR DESCRIPTION
**Pull request by JiangWeixian**
https://github.com/outline/rich-markdown-editor/issues/175

emoji

### CODE REVIEW

**Comment by tommoor** [comment]
Anything I can do to help here? Thanks~

**Comment by JiangWeixian** [comment]
> Anything I can do to help here? Thanks~

😂 sorry, I a little bit busy recently, this pr almost done, i will check some details as quick as possible

File: src/nodes/Blockquote.ts [hunk 1]
```diff
@@ -23,7 +23,10 @@ export default class Blockquote extends Node {
   }
 
   commands({ type }) {
-    return () => toggleWrap(type);
```
[the remark this answers is not in the document]
Comment (JiangWeixian) [reply src/nodes/Blockquote.ts:26]: yep, you are correct. It should be removed

File: src/components/EmojiMenuItem.tsx [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    [selected]
+  );
+
+  return (
+    <MenuItem
+      selected={selected}
+      onClick={disabled ? undefined : onClick}
+      ref={ref}
+    >
+      &nbsp;&nbsp;{title}
```
[the remark this answers is not in the document]
Comment (JiangWeixian) [reply src/components/EmojiMenuItem.tsx:40]: ok~

File: src/components/EmojiMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+      <Portal>
+        <Wrapper
+          id="at-menu-container"
+          active={isActive}
+          ref={this.menuRef}
+          {...positioning}
+        >
+          <List>
+            {items.map((item, index) => {
+              if (item.name === "separator") {
```
[the remark this answers is not in the document]
Comment (JiangWeixian) [reply src/components/EmojiMenu.tsx:311]: Thx, i will make an abstraction to reused code between emoji and block

File: src/index.tsx [hunk 4]
```diff
@@ -314,6 +318,10 @@ class RichMarkdownEditor extends React.PureComponent<Props, State> {
             dictionary,
             onShowToast: this.props.onShowToast,
           }),
+          new Emoji({
```
Comment (poitch) [inline src/index.tsx:321]: You will need add emoji to disableExtensions in props.

File: src/plugins/EmojiTrigger.tsx [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    return [
+      // main regex should match only:
+      // /word
+      new InputRule(OPEN_REGEX, (state, match) => {
+        console.log(
+          match &&
+            state.selection.$from.parent.type.name === "paragraph" &&
+            !isInTable(state),
+          match
+        );
```
Comment (poitch) [inline src/plugins/EmojiTrigger.tsx:101]: Another debug statement left behind

File: src/index.tsx (hunk 4 above)
Comment (JiangWeixian) [reply src/index.tsx:321]: thx, i will~

File: src/plugins/EmojiTrigger.tsx [hunk 6]
```diff
@@ -0,0 +1,92 @@
+import { InputRule } from "prosemirror-inputrules";
```
Comment (JiangWeixian) [inline src/plugins/EmojiTrigger.tsx:1]: BlockMenuTrigger and EmojiTrigger has lot of copy/paste code, but i not sure to create a common plugin or common trigger extension

File: src/components/BlockMenuItem.tsx [hunk 7]
```diff
@@ -47,13 +49,15 @@ function BlockMenuItem({
...
+      {Icon && (
+        <Icon
+          color={
+            selected ? theme.blockToolbarIconSelected : theme.blockToolbarIcon
+          }
+        />
+      )}
       &nbsp;&nbsp;{title}
-      <Shortcut>{shortcut}</Shortcut>
+      {shortcut && <Shortcut>{shortcut}</Shortcut>}
```
Comment (JiangWeixian) [inline src/components/BlockMenuItem.tsx:60]: make icon shortcut optional to let emojiitem resue blockmenuitem

File: src/components/BlockMenu.tsx [hunk 8]
```diff
@@ -1,626 +1,48 @@
...
-type State = {
-  insertItem?: EmbedDescriptor;
-  left?: number;
-  top?: number;
-  bottom?: number;
-  isAbove: boolean;
-  selectedIndex: number;
-};
-
-class BlockMenu extends React.Component<Props, State> {
```
Comment (JiangWeixian) [inline src/components/BlockMenu.tsx:50]: EmojiMenu BlockMenu use `CommonBlockMenu`

File: src/components/CommonBlockMenu.tsx [hunk 9]
```diff
@@ -0,0 +1,634 @@
...
+  embeds?: EmbedDescriptor[];
+  renderMenuItem: (
+    item: T,
+    index: number,
+    options: {
+      selected: boolean;
+      onClick: () => void;
+    }
+  ) => React.ReactNode;
+  filterable?: boolean;
```
Comment (JiangWeixian) [inline src/components/CommonBlockMenu.tsx:46]: filter items by search

File: src/components/CommonBlockMenu.tsx [hunk 10]
```diff
@@ -0,0 +1,634 @@
...
+  items: T[];
```
Comment (JiangWeixian) [inline src/components/CommonBlockMenu.tsx:47]: render menu item list

File: src/components/CommonBlockMenu.tsx [hunk 11]
```diff
@@ -0,0 +1,634 @@
...
+  dictionary: typeof baseDictionary;
+  view: EditorView;
+  search: string;
+  uploadImage?: (file: File) => Promise<string>;
+  onImageUploadStart?: () => void;
+  onImageUploadStop?: () => void;
+  onShowToast?: (message: string, id: string) => void;
+  onLinkToolbarOpen?: () => void;
+  onClose: () => void;
+  onClearSearch: () => void;
```
Comment (JiangWeixian) [inline src/components/CommonBlockMenu.tsx:36]: emojimenu and blockmenu has different clearsearch behavior

File: src/components/CommonBlockMenu.tsx [hunk 12]
```diff
@@ -0,0 +1,637 @@
...
+    } catch (err) {
+      console.warn(err);
+      return defaultPosition;
+    }
+
+    const domAtPos = view.domAtPos.bind(view);
+
+    const ref = this.menuRef.current;
+    const offsetHeight = ref ? ref.offsetHeight : 0;
+    const node = findDomRefAtPos(selection.from, domAtPos);
```
Comment (JiangWeixian) [inline src/components/CommonBlockMenu.tsx:358]: emoji should allow insert inside para

File: src/components/EmojiMenuItem.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (JiangWeixian) [reply src/components/EmojiMenuItem.tsx]: now, it look like this image

File: package.json [hunk 13]
```diff
@@ -26,8 +26,12 @@
   },
   "dependencies": {
     "copy-to-clipboard": "^3.0.8",
+    "fuzzy-search": "^3.2.1",
+    "gemoji": "^7.0.0",
+    "install": "^0.13.0",
```
Comment (tommoor) [inline package.json:31]:
```suggestion
```

Not needed, right?

File: src/nodes/Blockquote.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (tommoor) [reply src/nodes/Blockquote.ts:26]: I'm not sure why this is deleted… should remain.

File: src/nodes/Emoji.tsx [hunk 14]
```diff
@@ -0,0 +1,89 @@
+import { InputRule } from "prosemirror-inputrules";
+import { nameToEmoji } from "gemoji";
+import Node from "./Node";
+export default class Emoji extends Node {
```
Comment (tommoor) [inline src/nodes/Emoji.tsx:4]:
```suggestion

export default class Emoji extends Node {
```

File: src/plugins/EmojiTrigger.tsx [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+import { Plugin } from "prosemirror-state";
+import Extension from "../lib/Extension";
+import { run } from "./BlockMenuTrigger";
+
+const OPEN_REGEX = /:([0-9a-zA-Z_]+)?$/;
```
Comment (tommoor) [inline src/plugins/EmojiTrigger.tsx:6]: I think we need to account for ":+" and ":-" starts in the regex here for thumbs up/down

File: src/nodes/Emoji.tsx [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+          );
+        }
+
+        return tr;
+      }),
+    ];
+  }
+
+  toMarkdown(state, node) {
+    state.write(":" + (node.attrs["data-name"] || "dog") + ":");
```
Comment (tommoor) [inline src/nodes/Emoji.tsx:78]:
```suggestion
    const name = node.attrs["data-name"];
    if (name) {
      state.write(`:${name}:`);
    }
```

Probably best not to default to dog

File: src/components/EmojiMenuItem.tsx [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+const EmojiTitle = ({
+  emoji,
+  title,
+}: {
+  emoji: React.ReactNode;
+  title: React.ReactNode;
+}) => {
+  return (
+    <Title>
+      {emoji}
```
Comment (tommoor) [inline src/components/EmojiMenuItem.tsx:21]: Can we wrap the emoji to give it 16px size but keep the text size the same as the other menu?

File: src/components/EmojiMenuItem.tsx [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+function EmojiMenuItem(props: EmojiMenuItemProps) {
+  return (
+    <BlockMenuItem
+      {...props}
+      title={<EmojiTitle emoji={props.emoji} title={props.title} />}
+    />
+  );
+}
+
+export default withTheme(EmojiMenuItem);
```
Comment (tommoor) [inline src/components/EmojiMenuItem.tsx:41]:
```suggestion
export default EmojiMenuItem;
```

Shouldn't need `withTheme` here

File: src/components/EmojiMenu.tsx [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+  get items(): Emoji[] {
+    const { search = "" } = this.props;
+
+    const n = search.toLowerCase();
+    const result = searcher.search(n).map(item => {
+      const name = item.names[0];
+      return {
+        ...item,
+        name: "emoji",
+        title: name,
```
Comment (tommoor) [inline src/components/EmojiMenu.tsx:39]: Is it possible to use "description" from gemoji here so that the text doesn't contain "_"?

**Review by tommoor — COMMENTED** [review]
This is great – getting so close!

File: package.json (hunk 13 above)
Comment (JiangWeixian) [reply package.json:31]: my fault

File: src/components/EmojiMenu.tsx (hunk 19 above)
Comment (JiangWeixian) [reply src/components/EmojiMenu.tsx:39]: maybe not.. gemoji description longer than name

```json
{
  "emoji": "😍",
  "names": ["heart_eyes"],
  "tags": ["love", "crush"],
  "description": "smiling face with heart-eyes",
  "category": "Smileys & Emotion"
}
```

Comment (tommoor) [reply src/components/EmojiMenu.tsx:39]: Mm, maybe we can humanize the name then? (Capitalize, remove underscores)

Feels very strange to show underscores in user facing UI. I know GitHub does it, but GitHub is weird.

File: src/components/EmojiMenuItem.tsx (hunk 17 above)
Comment (JiangWeixian) [reply src/components/EmojiMenuItem.tsx:21]: ok ~

File: src/components/EmojiMenu.tsx (hunk 19 above)
Comment (JiangWeixian) [reply src/components/EmojiMenu.tsx:39]: maybe you are right, i will try it later.

Comment (JiangWeixian) [reply src/components/EmojiMenu.tsx:39]: image

**Comment by tommoor** [comment]
Hey hey, just testing out the branch and when I go to insert an emoji it always places the text `undefined` instead…

**Comment by JiangWeixian** [comment]
> Hey hey, just testing out the branch and when I go to insert an emoji it always places the text `undefined` instead…


https://user-images.githubusercontent.com/6839576/132935055-70351d0f-2ddd-4155-a080-45eec7e6ab83.mp4

**Comment by tommoor** [comment]
I had to make a couple of local fixes, but squashed and merged here – bbaf1574d2680e9eedec95d9afee77f849777bdd 

Thank you!

**Comment by tommoor** [comment]
Unfortunately the published version with this merged doesn't seem to work because of gemoji – I haven't yet figured out how to work around it.

**Comment by JiangWeixian** [comment]
😓 it looks require esm not supported, i will try changed tsconfig.json: `module` https://www.typescriptlang.org/tsconfig#es6

**Comment by JiangWeixian** [comment]
its wired, i try it on my webpack project, it works fine

**Comment by tommoor** [comment]
What did you change? Changing `module: es6` here prevents the lib from building

**Comment by JiangWeixian** [comment]
> What did you change? Changing `module: es6` here prevents the lib from building

i did not change yet, just download `v11.18.0`, there is mini app https://github.com/JiangWeixian/esm-app

**Comment by tommoor** [comment]
Ah, okay. It no longer works in `outline`, so it likely won't work in many other projects.

**Comment by JiangWeixian** [comment]
> Ah, okay. It no longer works in `outline`, so it likely won't work in many other projects.

I try it on `react-scripts`, it works fine. Is there any mini repo contain this error 🙈.

**Comment by tommoor** [comment]
I'm not sure outline is not a complicated build system – https://github.com/outline/outline/blob/main/server/.babelrc

I pretty much lost my whole evening to this, so signing off for now.

**Comment by JiangWeixian** [comment]
(Maybe) Solution 1

convert outline server to native esm

Solution 2

I lock gemoji version to cjs version https://github.com/outline/rich-markdown-editor/pull/555

**Comment by guatedude2** [comment]
Any idea why I'm getting this error when I run tests within my project?

```
 FAIL  apps/registration/src/legacy/components/WelcomeText/WelcomeText.test.tsx
  ● Test suite failed to run

    Cannot find module 'gemoji/name-to-emoji' from 'node_modules/rich-markdown-editor/dist/nodes/Emoji.js'

    Require stack:
      node_modules/rich-markdown-editor/dist/nodes/Emoji.js
      node_modules/rich-markdown-editor/dist/index.js
      packages/ui/src/components/content/RichContentEditor/BaseEditor.tsx
      packages/ui/src/components/content/Markdown/Markdown.tsx
      packages/ui/src/components/content/Markdown/index.ts
      packages/ui/src/components/index.ts
      apps/registration/src/legacy/components/WelcomeText/WelcomeText.tsx
      apps/registration/src/legacy/components/WelcomeText/index.ts
      apps/registration/src/legacy/components/WelcomeText/WelcomeText.test.tsx

      at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:306:11)
      at Object.<anonymous> (node_modules/rich-markdown-editor/src/nodes/Emoji.tsx:3:1)
```

**Comment by JiangWeixian** [comment]
@tommoor @guatedude2 took look at this pr, fix this problem....

**Comment by guatedude2** [comment]
> @tommoor @guatedude2 took look at this pr #557, fix this issue....

Thanks @JiangWeixian I'll keep an eye out on the PR!

**Pull request closed without merging**